lrc/Pos/Lrc/Core.hs

```diff
@@ -0,0 +1,80 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- | Core logic of LRC.
+
+module Pos.Lrc.Core
+       ( findDelegationStakes
+       , findRichmenStake
+       ) where
+
+import qualified Data.HashMap.Strict as HM
+import qualified Data.HashSet        as HS
+import           Universum
+
+import           Pos.Core.Coin       (mkCoin, unsafeAddCoin)
+import           Pos.Core.Types      (Coin, StakeholderId)
+import           Pos.Lrc.Types       (RichmenSet, RichmenStake)
+import           Pos.Util.Iterator   (MonadIterator (..))
+import           Pos.Util.Util       (getKeys)
+
+
+-- | Function helper for delegated richmen.
+-- Iterate by Delegate -> [Issuer] map and compute two set:
+-- 1. Old richmen who delegated own stake and isn't richman more.
+-- 2. Delegates who became richmen.
+findDelegationStakes
+    :: forall m . MonadIterator (StakeholderId, [StakeholderId]) m
+    => (StakeholderId -> m Bool) -- helper
+    -> (StakeholderId -> m (Maybe Coin)) -- helper
+    -> Coin
+    -> m (RichmenSet, RichmenStake) -- old richmen, new richmen
+findDelegationStakes isIssuer stakeResolver t = do
+    (old, new) <- step (mempty, mempty)
+    pure (getKeys ((HS.toMap old) `HM.difference` new), new)
+  where
+    step :: (RichmenSet, RichmenStake)
+         -> m (RichmenSet, RichmenStake)
+    step richmen = nextItem @(StakeholderId, [StakeholderId]) >>=
+        maybe (pure richmen) (onItem richmen >=> step)
+    onItem (old, new) (delegate, issuers) = do
+        sumIssuers <-
+          foldM (\cr id -> (unsafeAddCoin cr) <$> safeBalance id)
```

Comment:
2 spaces. The whole `findDelegationStakes` after `where` is poorly indented/barely readable.